ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java

```diff
@@ -101,4 +205,468 @@ public Integer getValue() {
+          registry.counter(MetricRegistry.name(NonBlockingRouter.class, dataNodeName, "-UnknownRouterErrorForRequest"));
+
+      // Misc metrics at the DataNode level.
+      requestErrorRate = metricRegistry.meter(MetricRegistry.name(NonBlockingRouter.class, "-RequestErrorArrivalRate"));
+      unknownReplicaResponseError =
+          registry.counter(MetricRegistry.name(NonBlockingRouter.class, dataNodeName, "-UnknownReplicaResponseError"));
+    }
+
+    /**
+     * Update the request arrival rate for a given {@link com.github.ambry.router.NonBlockingRouter.RouterOperationType}.
+     * This method should be called when a request to a {@link DataNodeId} is created.
+     */
+    void uopdateRequestRate(NonBlockingRouter.RouterOperationType operationType) {
+      switch (operationType) {
+        case PutBlob:
+          putRequestRate.mark();
+          break;
+        case GetBlobInfo:
+          getBlobInfoRequestRate.mark();
+          break;
+        case GetBlob:
+          getRequestRate.mark();
+          break;
+        case DeleteBlob:
+          deleteRequestRate.mark();
+          break;
+        default:
+          // We simply log the error here and do not add a metrics to record this, as the default case should not happen.
+          logger.warn("Error for unknown RouterOperationType for counting request rate: " + operationType);
+          break;
+      }
+    }
+
+    /**
+     * Update request latency for a given {@link com.github.ambry.router.NonBlockingRouter.RouterOperationType}.
+     * This method should be called when a response of a request has been received and processed. This method should
+     * not be called when a request is aborted.
+     * @param latency Request latency.
+     * @param operationType {@link com.github.ambry.router.NonBlockingRouter.RouterOperationType}.
+     */
+    void updateRequestLatency(long latency, NonBlockingRouter.RouterOperationType operationType) {
+      switch (operationType) {
+        case PutBlob:
+          putRequestLatencyMs.update(latency);
+          break;
+        case GetBlobInfo:
+          getBlobInfoRequestLatencyMs.update(latency);
+          break;
+        case GetBlob:
+          getRequestLatencyMs.update(latency);
+          break;
+        case DeleteBlob:
+          deleteRequestLatencyMs.update(latency);
+          break;
+        default:
+          // We simply log the error here and do not add a metrics to record this, as the default case should not happen.
+          logger.warn("Error for unknown RouterOperationType for updating request latency: " + operationType);
+          break;
+      }
+    }
+
+    /**
+     * Update the counter for timedout requests for a given {@link com.github.ambry.router.NonBlockingRouter.RouterOperationType}.
+     * @param operationType {@link com.github.ambry.router.NonBlockingRouter.RouterOperationType}
+     */
+    void updateRequestTimeoutCount(NonBlockingRouter.RouterOperationType operationType) {
```

Comment:
Unnecessary. We should try not to overcomplicate stuff. Also, I think you are being influenced by the coordinator code to some extent.